crates/llm-ls/src/adaptors.rs

```diff
@@ -0,0 +1,163 @@
+    );
+
+    if let Some(api_token) = api_token {
+        headers.insert(
+            AUTHORIZATION,
+            HeaderValue::from_str(&format!("Bearer {api_token}")).map_err(internal_error)?,
+        );
+    }
+
+    Ok(headers)
+}
+
+fn parse_tgi_text(text: reqwest::Result<String>) -> Result<Vec<Generation>, jsonrpc::Error> {
+    let generations =
+        match serde_json::from_str(&text.unwrap_or_default()).map_err(internal_error)? {
+            APIResponse::Generation(gen) => vec![gen],
+            APIResponse::Generations(_) => {
+                return Err(internal_error(
+                    "TGI parser unexpectedly encountered api-inference",
+                ))
+            }
+            APIResponse::Error(err) => return Err(internal_error(err)),
+        };
+    Ok(generations)
+}
+
+fn build_api_body(prompt: String, params: CompletionParams) -> Value {
+    build_tgi_body(prompt, params)
+}
+
+fn build_api_headers(api_token: Option<&String>, ide: Ide) -> Result<HeaderMap, jsonrpc::Error> {
+    build_tgi_headers(api_token, ide)
+}
+
+fn parse_api_text(text: reqwest::Result<String>) -> Result<Vec<Generation>, jsonrpc::Error> {
+    let generations =
+        match serde_json::from_str(&text.unwrap_or_default()).map_err(internal_error)? {
+            APIResponse::Generation(gen) => vec![gen],
+            APIResponse::Generations(gens) => gens,
+            APIResponse::Error(err) => return Err(internal_error(err)),
+        };
+    Ok(generations)
+}
+
+fn build_ollama_body(prompt: String, params: CompletionParams) -> Value {
+    let request_body = params.request_body.unwrap_or_default();
+    let body = serde_json::json!({
+        "prompt": prompt,
+        "model": request_body.get("model"),
+    });
+    body
+}
+fn build_ollama_headers() -> Result<HeaderMap, jsonrpc::Error> {
+    let headers = HeaderMap::new();
+    Ok(headers)
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct OllamaGeneration {
+    response: String,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+enum OllamaAPIResponse {
+    Generation(OllamaGeneration),
+    Error(APIError),
+}
+
+fn parse_ollama_text(
+    text: Result<String, reqwest::Error>,
+) -> Result<Vec<Generation>, jsonrpc::Error> {
+    match text {
+        Ok(text) => {
+            let mut gen: Vec<Generation> = Vec::new();
+            for row in text.split('\n') {
+                if row.is_empty() {
+                    continue;
+                }
+                let chunk = match serde_json::from_str(row) {
+                    Ok(OllamaAPIResponse::Generation(ollama_gen)) => ollama_gen.response,
+                    Ok(OllamaAPIResponse::Error(err)) => return Err(internal_error(err)),
+                    Err(err) => return Err(internal_error(err)),
+                };
+                gen.push(Generation {
+                    generated_text: chunk,
+                })
+            }
+            Ok(gen)
```

Comment:
This will not work as by default ollama streams chunks of the generation one at a time. The way you're currently doing this will result in showing only the first chunk to the user as a suggestion.
Say the prompt is:
```py
def add(a, b):\n\t
```
the response from ollama would look like this:
```json
{ "model": "llama2", "created_at": "2023-08-04T08:52:19.385406455-07:00", "response": "a +", "done": false }
{ "model": "llama2", "created_at": "2023-08-04T08:52:19.385406455-07:00", "response": "b", "done": false }
{ "model": "llama2", "created_at": "2023-08-04T19:22:45.499127Z", "response": "", "context": [1, 2, 3], "done": true, "total_duration": 5589157167, "load_duration": 3013701500, "sample_count": 114, "sample_duration": 81442000, "prompt_eval_count": 46, "prompt_eval_duration": 1160282000, "eval_count": 113, "eval_duration": 1325948000 }
```
Thus only suggesting to the user `a +` when we wanted to output `a + b`.

I suggested to add `stream: false` to the body of the request, this way there will only be one JSON response. I think this will be simpler than having to build the generation from each chunk, but up to you.

Reply:
Good find, this seems to be a recent addition but more than happy to add in since it will certainly make things simpler